Add a 404 page for the application

sewagodimo opened the pull request:
The 404 landing page, this makes user testing much simpler.
It's kinda awkward when students run unit a 404 Server Error

sewagodimo committed:
Add a view for 404 not found page

sewagodimo committed:
update tests for the 404 error page

PiDelport reviewed (commented):
We can probably do this without any custom view or such, and just a usual `404.html` template?

PiDelport committed:
(Dev settings: Comment ALLOWED_HOSTS)

PiDelport committed:
Switch to using Django's built-in 404 view

This uses the default "404.html" template relative to the base, and
returns a 404 Not Found status code, rather than 200 OK.

PiDelport commented on src/buza/templates/404.html:
```diff
@@ -0,0 +1,11 @@
+{% block title %}Page not found  404 {{ block.super }}{% endblock %}
+{% block head_extra %}
+    {{ block.super }}
+    <link rel="stylesheet" href="{% static 'buza/css/errors.css' %}">
```
@sewagodimo: This `errors.css` file seems to be missing: is this one you added locally, but haven't committed yet?

sewagodimo replied:
No, I didn't realise that was still there
I want to style the 404 error page and the 400 error pages but it'snot a priority for now

sewagodimo commented on tests/test_views.py:
```diff
@@ -630,3 +631,16 @@ def test_get(self) -> None:
         self.assertContains(response, question.title, count=1)
+
+
+class Test404PageNotFound(TestCase):
```
@pjdelport I didn't know this, all the tutorials kept referring to it :)
Thank you the heads up

PiDelport reviewed (changes requested):
@sewagodimo: Okay, I made the changes to switch this to using Django's built-in 404 view: all we need for that to work is to name our template `404.html`, without the prefix.

Can you look at that comment about `buza/css/errors.css`: is that something you still want to add?

sewagodimo commented on src/buza/settings_env_dev.py:
```diff
@@ -27,4 +27,7 @@
 ]
 
+# Include the local host by default for development.
+ALLOWED_HOSTS = ['127.0.0.1', 'localhost']
```
@pjdelport do we still need this then?

PiDelport replied:
It's not a bad idea to keep: it's not used when `DEBUG` is enabled, but it's useful to have for testing / developing with `DEBUG` off (such as when testing the 404 template).

sewagodimo commented:
@pjdelport  I am happy if you merge, when travis is happy

sewagodimo committed:
remove errors.css

We many need it in future to style the 404 error page but it's not a priority for now

PiDelport committed:
Tweak 404 page content a bit

Now with more emoji!

PiDelport commented:
@sewagodimo: I think Travis got wedged somehow when I rebased and pushed this… I'm not sure how to get it unwedged. :/ Maybe it's because GitHub's current systems being partially down?

We can probably go ahead and merge it: the tests do pass.

I made a last tweak to the 404 page content: do you want to check it out and see if you're happy? If you are, go ahead and merge! :coffee:

sewagodimo commented:
@pjdelport awesome, if I had known that you were free tonight I would have come to your place, next time!

sewagodimo commented on src/buza/templates/404.html:
```diff
@@ -0,0 +1,11 @@
+{% block content %}
+    <div class="text-center">
+        <h1>We could not find the page you were looking for</h1>
+        <p style="font-size: 3rem">😭</p>
```
super cute

sewagodimo commented:
@pjdelport should we wait for the coverage report? It seems to be taking a while

PiDelport commented:
@sewagodimo: Well, the branch coverage and patch coverage passes so far. I think the PR build (and its coverage) may just be stalled due to GitHub's technical issues.

We could wait until tomorrow (Tuesday), to see if it actually gets triggered and builds, and then merge if it does?

PiDelport commented:
> @pjdelport awesome, if I had known that you were free tonight I would have come to your place, next time!

I was hacking from Codebridge tonight! :D
